src/sources/prometheus/remote_write.rs

```diff
@@ -0,0 +1,263 @@
+use crate::{
+    config::{self, GenerateConfig, GlobalOptions, SourceConfig, SourceDescription},
+    event::{Metric, MetricKind, MetricValue},
+    internal_events::{
+        PrometheusNoNameError, PrometheusRemoteWriteParseError, PrometheusRemoteWriteReceived,
+        PrometheusRemoteWriteSnapError,
+    },
+    prometheus::{proto, METRIC_NAME_LABEL},
+    shutdown::ShutdownSignal,
+    sources::{
+        self,
+        util::{ErrorMessage, HttpSource, HttpSourceAuthConfig},
+    },
+    tls::TlsConfig,
+    Event, Pipeline,
+};
+use bytes::Bytes;
+use chrono::{DateTime, TimeZone, Utc};
+use prost::Message;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{BTreeMap, HashMap},
+    net::SocketAddr,
+};
+use warp::http::{HeaderMap, StatusCode};
+
+const SOURCE_NAME: &str = "prometheus_remote_write";
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct PrometheusRemoteWriteConfig {
+    address: SocketAddr,
+
+    tls: Option<TlsConfig>,
+
+    auth: Option<HttpSourceAuthConfig>,
+}
+
+inventory::submit! {
+    SourceDescription::new::<PrometheusRemoteWriteConfig>(SOURCE_NAME)
+}
+
+impl GenerateConfig for PrometheusRemoteWriteConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            address: "127.0.0.1:9090".parse().unwrap(),
+            tls: None,
+            auth: None,
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "prometheus_remote_write")]
+impl SourceConfig for PrometheusRemoteWriteConfig {
+    async fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<sources::Source> {
+        let source = RemoteWriteSource {
+            decompressor: snap::raw::Decoder::new(),
+        };
+        source.run(self.address, "", &self.tls, &self.auth, out, shutdown)
+    }
+
+    fn output_type(&self) -> crate::config::DataType {
+        config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        SOURCE_NAME
+    }
+}
+
+#[derive(Clone)]
+struct RemoteWriteSource {
+    decompressor: snap::raw::Decoder,
+}
+
+impl RemoteWriteSource {
+    fn decode_body(&self, body: Bytes) -> Result<Vec<Event>, ErrorMessage> {
+        let body = self
+            .decompressor
+            .clone()
```

Comment:
Tiny point, I don't think this `clone` is necessary. Looking at the code it doesn't look like `decompress_vec` changes anything in the struct, despite it being `mutable`. If it does need to be cloned then does it need to be a member of `RemoteWriteSource`, as you could just create a new instance each time?

The `Decoder` struct is so small it probably doesn't make any difference,  but might make the code cleaner.

Reply:
It does need the `clone` because `self` is not `mut` here, and so `self.compressor` isn't either. I can't make `decode_body` take `&mut self` since it's called from a non-mut context.